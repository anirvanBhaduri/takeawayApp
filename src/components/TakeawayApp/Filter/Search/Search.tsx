import * as React from 'react';

import {Filter, FilterProps} from "../Filter";
import {RestaurantProps} from "../../RestaurantList/Restaurant/Restaurant";

/**
 * The searchTerm is required to ensure
 * we can search.
 *
 * @interface SearchProps
 */
interface SearchProps extends FilterProps {
    searchTerm: string;
}

/**
 * Search using a string.
 *
 * Used to filter RestaurantProps.
 *
 * @class Search
 * @package TakeawayApp
 *
 * @author Anirvan Bhaduri
 * @since 9th Dec 2018
 * @version 0.0.1
 */
export class Search extends Filter<SearchProps> {

    /**
     * The search term used to filter the results.
     */
    searchTerm: string;

    /**
     * Search constructor.
     *
     * @param props
     */
    constructor(props: any) {
        super(props);

        this.searchTerm = this.props.searchTerm;
    }

    /**
     * Set the search term.
     *
     * @param searchTerm
     */
    setSearchTerm(searchTerm: string) {
        this.searchTerm = searchTerm;
    }

    /**
     * Get the current search term.
     *
     * @return string
     */
    getSearchTerm(): string {
        return this.searchTerm;
    }

    /**
     * Apply the search filter.
     *
     * @param list
     *
     * @return RestaurantProps
     */
    applyFilter(list: RestaurantProps[]): RestaurantProps[] {
        // Do some filter logic here using searchTerm
        // ... should mainly remove elements

        // When we don't have a search term, we don't want to filter at all
        if (!this.searchTerm) {
            return list;
        }

        const searchTerm = this.searchTerm.toLowerCase();
        return list.filter(
            (value: RestaurantProps): boolean => {
                return value.name.toLowerCase().indexOf(searchTerm) >= 0;
            }
        );
    }

    /**
     * Render the search component.
     *
     * @return JSX
     */
    render() {
        const handler = this.props.filterHandler;
        const searchTerm = this.props.searchTerm;

        return (
            <div>
                <input
                    onChange={handler}
                    value={searchTerm}
                    type="input"
                />
            </div>
        );
    }
}
