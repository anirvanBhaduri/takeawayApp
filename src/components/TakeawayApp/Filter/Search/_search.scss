/// Search component styles.
///
/// @author Anirvan Bhaduri
/// @since 10th Dec 2018

@import '../../../UI/_colour_constants.scss';
@import '../../../UI/size_constants';

// The search bar's styles
.search {
  float: left;
  width: 550px;

  @media screen and (max-width: $screen-break-medium) {
    width: 98%;
    margin: 10px;
  }

  @media screen and (max-width: $screen-break-small) {
    width: 98%;
    margin: 10px;
  }
}

// The search bar's input's styles
.search__input {
  display: inline-block;
  @include input-styles(28px, 545px);

  @media screen and (max-width: $screen-break-medium) {
    font-size: inherit;
  }

  @media screen and (max-width: $screen-break-small) {
    font-size: inherit;
  }
}