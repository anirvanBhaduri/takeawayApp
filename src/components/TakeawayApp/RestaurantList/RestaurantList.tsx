import * as React from 'react';

import {Restaurant, RestaurantProps} from "./Restaurant/Restaurant";
import {sortables} from "../Filter/Sort/Sortables";

/**
 * Defines the format for a RestaurantList object.
 *
 * @interface RestaurantListProps
 *
 * @property array<Restaurant> An array containing only Restaurants.
 */
export interface RestaurantListProps {
    restaurants: [];
    filterCriteria?: {
        [index: string]: any;
    };
}

/**
 * The RestaurantList is responsible for arranging
 * a list of restaurants based on a sort criteria.
 *
 * @class RestaurantList
 *
 * @author Anirvan Bhaduri
 * @since 9th Dec 2018
 * @version 0.0.1
 */
export class RestaurantList extends React.Component<RestaurantListProps, {}> {

    /**
     * Initialise state with restaurants.
     *
     * @property state
     */
    state: {
        restaurants: [],
    };

    /**
     * RestaurantList constructor.
     *
     * Load in the restaurants here.
     *
     * @param props
     */
    constructor(props: any) {
        super(props);

        // Initialise state.
        this.state = {
            restaurants: this.props.restaurants,
        };

        // Bind event handlers.
        this.handleFavouriteClick = this.handleFavouriteClick.bind(this);
    }

    /**
     * Handle the event when someone clicks the favourite
     * button for a restaurant.
     *
     * @param e
     */
    handleFavouriteClick(e: any) {
        const restaurantName = e.target.value;
        const {restaurants} = this.state;

        restaurants.map((restaurant: RestaurantProps) => {
            if (restaurant.name !== restaurantName) {
                return restaurant;
            }

            // We can now be sure the name is the same
            if (restaurant.favourite === null) {
                restaurant.favourite = true;
            } else {
                // We invert the value if its already set.
                restaurant.favourite = !restaurant.favourite;
            }
        });

        this.setState({restaurants: restaurants});
    }

    /**
     * Filter the restaurant list using the filter criteria
     * passed to this object.
     *
     * @return JSX.Element[]
     */
    filter(
        restaurants: [],
        criteria: {[index: string]: any}
    ): JSX.Element[] {
        let orderedRestaurants = restaurants;

        Object.keys(criteria).map((key: string): void => {
            orderedRestaurants = criteria[key].applyFilter(orderedRestaurants);
        });

        const sortType = criteria.sort.getSortType();

        return orderedRestaurants.map((props: RestaurantProps) => {
            return <Restaurant
                key={props.name}
                sortValue={sortables[sortType].sortValue(props, sortType)}
                sortName={sortables[sortType].title}
                clickHandler={this.handleFavouriteClick}
                favourite={props.favourite}
                name={props.name}
                status={props.status}
            />;
        });
    }

    /**
     * Render the component.
     *
     * @return JSX
     */
    render() {
        const restaurants = this.filter(
            this.state.restaurants,
            this.props.filterCriteria
        );

        return (
            <div className={'restaurant-list'} >
                {restaurants}
            </div>
        )
    }
}
