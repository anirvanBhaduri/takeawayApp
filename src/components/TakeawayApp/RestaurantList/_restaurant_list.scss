/// Restaurant List component styles.
///
/// @author Anirvan Bhaduri
/// @since 10th Dec 2018

@import '../../UI/_colour_constants.scss';
@import '../../UI/size_constants';

// The Restaurant List's styles
.restaurant-list {
  margin-top: 10px;
  padding: 0.1em;
  width: 900px;
  display: inline-block;
  overflow: hidden;
  overflow-y: scroll;
  height: $app-body-height - 100;
}