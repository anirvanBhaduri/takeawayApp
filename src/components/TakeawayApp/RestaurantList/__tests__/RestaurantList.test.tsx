import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { shallow } from 'enzyme';

// SUT
import {RestaurantList} from '../RestaurantList';
import {Search} from "../../Filter/Search/Search";

/**
 * Load data to use on the SUT.
 */
const loadedData = require('../../../../../dist/Sample.json');

/**
 * RestaurantList component tests.
 *
 * @covers RestaurantList
 */
describe('RestaurantList Component Tests', () => {

    /**
     * RestaurantList renders correctly.
     *
     * @covers RestaurantList.render
     */
    it('should render correct snapshot', () => {
        const {restaurants} = loadedData;
        const filterCriteria = {
            search: new Search({
                searchTerm: '',
                filterHandler: () => {}
            })
        };

        const wrapper = shallow(
            <RestaurantList
                restaurants={restaurants}
                filterCriteria={filterCriteria}
            />
        );

        expect(wrapper).toMatchSnapshot();
    })
});
