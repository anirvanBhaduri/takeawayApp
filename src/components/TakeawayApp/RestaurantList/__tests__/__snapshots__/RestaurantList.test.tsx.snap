// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RestaurantList Component Tests should render correct snapshot of RestaurantList 1`] = `
<div
  className="restaurant-list"
>
  <Component
    clickHandler={[Function]}
    sortName="Best Match"
    sortValue={1}
  />
</div>
`;
